smartsim/entity/application.py

```diff
@@ -0,0 +1,300 @@
+        """The executable to run"""
+        self._exe_args = self._build_exe_args(exe_args) or []
+        """The executable arguments"""
+        self._files = copy.deepcopy(files) if files else None
+        """Files to be copied, symlinked, and/or configured prior to execution"""
+        self._file_parameters = (
+            copy.deepcopy(file_parameters) if file_parameters else {}
+        )
+        """Parameters and values to be used when configuring files"""
+        self._incoming_entities: t.List[SmartSimEntity] = []
+        """Entities for which the prefix will have to be known by other entities"""
+        self._key_prefixing_enabled = False
+        """Unique prefix to avoid key collisions"""
+
+    @property
+    def exe(self) -> str:
+        """Return executable to run.
+
+        :returns: application executable to run
+        """
+        return self._exe
+
+    @exe.setter
+    def exe(self, value: str) -> None:
+        """Set executable to run.
+
+        :param value: executable to run
+        """
+        self._exe = copy.deepcopy(value)
+
+    @property
+    def exe_args(self) -> t.Sequence[str]:
+        """Return a list of attached executable arguments.
+
+        :returns: application executable arguments
+        """
+        return self._exe_args
+
+    @exe_args.setter
+    def exe_args(self, value: t.Union[str, t.Sequence[str], None]) -> None:  #
+        """Set the executable arguments.
+
+        :param value: executable arguments
+        """
+        self._exe_args = self._build_exe_args(value)
+
+    @property
+    def files(self) -> t.Optional[EntityFiles]:
+        """Return files to be copied, symlinked, and/or configured prior to
+        execution.
+
+        :returns: files
+        """
+        return self._files
+
+    @files.setter
+    def files(self, value: t.Optional[EntityFiles]) -> None:
+        """Set files to be copied, symlinked, and/or configured prior to
+        execution.
+
+        :param value: files
+        """
+        self._files = copy.deepcopy(value)
+
+    @property
+    def file_parameters(self) -> t.Mapping[str, str]:
+        """Return file parameters.
+
+        :returns: application file parameters
+        """
+        return self._file_parameters
+
+    @file_parameters.setter
+    def file_parameters(self, value: t.Mapping[str, str]) -> None:
+        """Set the file parameters.
+
+        :param value: file parameters
+        """
+        self._file_parameters = copy.deepcopy(value)
+
+    @property
+    def incoming_entities(self) -> t.List[SmartSimEntity]:
+        """Return incoming entities.
+
+        :returns: incoming entities
+        """
+        return self._incoming_entities
+
+    @incoming_entities.setter
+    def incoming_entities(self, value: t.List[SmartSimEntity]) -> None:
+        """Set the incoming entities.
+
+        :param value: incoming entities
+        """
+        self._incoming_entities = copy.deepcopy(value)
```

Comment:
I know the whole idea of "key prefixing" and "incoming entities" is VERY up in the air at the moment, but I think it may actually make sense for this to be a shallow copy. Presumably if a user does:
```py
app_a = Application(...)
app_b = Application(...)

app_a.incoming_entities = [app_b]
exp = Experiment(...)

exp.start(Job(app_a, ...), Job(app_b, ...))
```
then `app_a` needs to know a reference to the `app_b` that is actually being launched, not a just a copy to some other application in memory that just happens to look like `app_b` at the time of registration.

Feel free to correct me on this though @amandarichardsonn @juliaputko @mellis13!

Reply:
I see!

Reply:
changed to shallow copy (copy.copy())